yt/utilities/lib/interpolators.pyx

```diff
@@ -110,6 +110,67 @@ def TrilinearlyInterpolate(np.ndarray[np.float64_t, ndim=3] table,
                    + table[x_i+1,y_i+1,z_i  ] * (xp*yp*zm) \
                    + table[x_i+1,y_i+1,z_i+1] * (xp*yp*zp)
 
+@cython.cdivision(True)
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def QuadrilinearlyInterpolate(np.ndarray[np.float64_t, ndim=4] table,
+                              np.ndarray[np.float64_t, ndim=1] x_vals,
+                              np.ndarray[np.float64_t, ndim=1] y_vals,
+                              np.ndarray[np.float64_t, ndim=1] z_vals,
+                              np.ndarray[np.float64_t, ndim=1] w_vals,
+                              np.ndarray[np.float64_t, ndim=1] x_bins,
+                              np.ndarray[np.float64_t, ndim=1] y_bins,
+                              np.ndarray[np.float64_t, ndim=1] z_bins,
+                              np.ndarray[np.float64_t, ndim=1] w_bins,
+                              np.ndarray[np.int_t, ndim=1] x_is,
+                              np.ndarray[np.int_t, ndim=1] y_is,
+                              np.ndarray[np.int_t, ndim=1] z_is,
+                              np.ndarray[np.int_t, ndim=1] w_is,
+                              np.ndarray[np.float64_t, ndim=1] output):
+    cdef double x, xp, xm
+    cdef double y, yp, ym
+    cdef double z, zp, zm
+    cdef double w, wp, wm
+    cdef double dx_inv, dy_inv, dz_inv, dw_inv
+    cdef int i, x_i, y_i, z_i, w_i
+    for i in range(x_vals.shape[0]):
+        x_i = x_is[i]
+        y_i = y_is[i]
+        z_i = z_is[i]
+        w_i = w_is[i]
+        x = x_vals[i]
+        y = y_vals[i]
+        z = z_vals[i]
+        w = w_vals[i]
+        dx_inv = 1.0 / (x_bins[x_i+1] - x_bins[x_i])
+        dy_inv = 1.0 / (y_bins[y_i+1] - y_bins[y_i])
+        dz_inv = 1.0 / (z_bins[z_i+1] - z_bins[z_i])
+        dw_inv = 1.0 / (w_bins[w_i+1] - w_bins[w_i])
```

Comment:
A note on performance: it might be less wasteful to store these intermediate values in a temporary array to prevent them from being recomputed over and over.

Reply:
Hm, that's the case if they're equally spaced bins, but do we mandate that?  They really might vary on the iteration.